### ISSUE TITLE
Location of read_data in partitioned-heat correct?

### ISSUE THREAD

**Issue by BenjaminRueth** [precice/tutorials#99]
Is the current location of read_data correct?

https://github.com/precice/tutorials/blob/b3f92fc54738fd1d006e9ee444d5127b2668cbd5/HT/partitioned-heat/fenics-fenics/heat.py#L206

Note that we initialize our data before that point

https://github.com/precice/tutorials/blob/b3f92fc54738fd1d006e9ee444d5127b2668cbd5/HT/partitioned-heat/fenics-fenics/heat.py#L124-L136

Therefore, I think that `read_data` should go after `advance` 

https://github.com/precice/tutorials/blob/b3f92fc54738fd1d006e9ee444d5127b2668cbd5/HT/partitioned-heat/fenics-fenics/heat.py#L225

maybe with an additional `is_coupling_ongoing` check.

@IshaanDesai, @uekerman Any opinions on this? Am I missing something?

**Comment by uekerman** [comment]
Yes, you could move `read_data` here after `advance`. I would not use `is_coupling_ongoing` there. If you want to prevent unnecessary reading you should use `is_read_data_available`, but that might be a bit inconsistent as you don't use `is_write_data_required` either. 

The inconsistency comes from the fact that `adapter.initialize_data` returns read data, because you need it for `create_coupling_expression`. As a user of the FEniCS adapter you need to either call `intitialize_data` or `read_data` before you can call `create_coupling_expression`, correct? What happens if `initialize_data` returns `None`? Will it still work? If not, not all combinations of serial / parallel coupling and initialize on / off will work.

**Comment by IshaanDesai** [comment]
Some points on why the `precice.read_data()` call was placed at the location it is now:
* Current location of `read_data()` call keeps the time loop consistent with the [Adapter Example](https://github.com/precice/precice/wiki/Adapter-Example#implicit-coupling) implicit coupling skeleton
* If `precice.initialize_data()` returns `None` then `create_coupling_expression` generates the FEniCS Expression by using zero valued data, hence the code will work. The `fenics-adapter` also throws [a warning](https://github.com/precice/fenics-adapter/blob/e951f25c8befa6c844c4e5f850ec417cfe4bc03a/fenicsadapter/fenicsadapter.py#L99) that no `inital_data` was provided.
* `read_data` acquired from `precice.read_data()` is used to update the coupling expression in `precice.update_coupling_expression(coupling_expression, read_data)`. If we move `read_data` call after `advance()` we will have to use `initial_data` to update the coupling expression in the first time window and then move to using the `read_data` generated. The current workflow avoids manual handling of these conditionals.

**Comment by uekerman** [comment]
What about the following:
* Let's not handle initial data in `create_coupling_expression`: let's remove the input argument of this function.
* Then, we also don't need `initialize_data` to return data. Actually, we could even move `initialize_data` into `initialize`. That would simplify much from user perspective I guess.
* Then, we could keep the `read_data` at the beginning of the while loop. It would then also handle the initial data.

?

Also, let's please remove the warning if no initial data is provided. That's a valid use case of preCICE.

**Comment by IshaanDesai** [comment]
>     * Let's not handle initial data in `create_coupling_expression`: let's remove the input argument of this function.
> 

I am not sure about this. Then do we always create a FEniCS Expression with empty data? This prevents the user from initializing the Expression to a non-standard initial state.

>     * Then, we also don't need `initialize_data` to return data. Actually, we could even move `initialize_data` into `initialize`. That would simplify much from user perspective I guess.
> 

In the initial design `initialize_data` was planned to be inside `initialize` itself. But then the user needs to provide the FEniCS function corresponding to the data being written in the coupling to `initialize` in addition to the already high number of input arguments. The thinking of having separate functions was to make it clear to the user that `initialize_data` is for a non-standard initialization and is in fact optional.

> 
> Also, let's please remove the warning if no initial data is provided. That's a valid use case of preCICE.

Sorry, here I am not sure I understand why we remove the warning because it is a valid use case.

**Comment by uekerman** [comment]
> I am not sure about this. Then do we always create a FEniCS Expression with empty data? This prevents the user from initializing the Expression to a non-standard initial state.

Is there a situation where this is needed? So, where not preCICE handles it, but the user wants to initialize sth else? It will be overwritten anyway in the first `update_coupling_expression`.

> > * Then, we also don't need `initialize_data` to return data. Actually, we could even move `initialize_data` into `initialize`. That would simplify much from user perspective I guess.
> In the initial design `initialize_data` was planned to be inside `initialize` itself. But then the user needs to provide the FEniCS function corresponding to the data being written in the coupling to `initialize` in addition to the already high number of input arguments. The thinking of having separate functions was to make it clear to the user that `initialize_data` is for a non-standard initialization and is in fact optional.

Yes, I see your point here.

> > Also, let's please remove the warning if no initial data is provided. That's a valid use case of preCICE.
> 
> Sorry, here I am not sure I understand why we remove the warning because it is a valid use case.

OK, I guess in the current situation the warning indeed makes sense. But no longer if we apply the suggested changes.
